### ISSUE TITLE
Java Build: nmake error

### ISSUE THREAD

**Issue by OisinMcArdle** [Z3Prover/z3#1275]
Hi I'm trying to build this for java.
Here's the commands I ran in Developer Command Prompt for VS 2017 and their output
[DeveloperCMDandOutput.txt](https://github.com/Z3Prover/z3/files/1342478/DeveloperCMDandOutput.txt)

Everything seems to work until I use 'nmake'. Googling the error isn't helping. I found a previous issue with this in here but it was closed without an answer.

Here's the errors when I run 'nmake' in the build directory:

C:\Z3Prover\z3\build>nmake

Microsoft (R) Program Maintenance Utility Version 14.11.25508.2
Copyright (C) Microsoft Corporation.  All rights reserved.

datalog_frontend.cpp
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winnt.h(21436): error C3861: '__readfsdword': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9044): error C2065: 'InterlockedIncrement64': undeclared identifier
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9076): error C2065: 'InterlockedDecrement64': undeclared identifier
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9112): error C3861: 'InterlockedExchange64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9166): error C3861: 'InterlockedExchangeAdd64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9176): error C3861: 'InterlockedExchangeAdd64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9223): error C3861: 'InterlockedAnd64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9233): error C3861: 'InterlockedOr64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winbase.h(9243): error C3861: 'InterlockedXor64': identifier not found
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.11.25503\bin\HostX86\x86\cl.EXE"' : return code '0x2'
Stop.

C:\Z3Prover\z3\build>

================

Trying what @delcypher recommended. It's also in the comments below but anyways:

cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON -->Tried this didn't work
Tried this command:
cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON --x64 --> THis also didn't work

Tried this command:
cmake -G -DBUILD_JAVA_BINDINGS=ON

Output:
CMake Error: Could not create named generator -DBUILD_JAVA_BINDINGS=ON

Generators
Visual Studio 15 2017 [arch] = Generates Visual Studio 2017 project files.
Optional [arch] can be "Win64" or "ARM".
Visual Studio 14 2015 [arch] = Generates Visual Studio 2015 project files.
Optional [arch] can be "Win64" or "ARM".
Visual Studio 12 2013 [arch] = Generates Visual Studio 2013 project files.
Optional [arch] can be "Win64" or "ARM".
Visual Studio 11 2012 [arch] = Generates Visual Studio 2012 project files.
Optional [arch] can be "Win64" or "ARM".
Visual Studio 10 2010 [arch] = Generates Visual Studio 2010 project files.
Optional [arch] can be "Win64" or "IA64".
Visual Studio 9 2008 [arch] = Generates Visual Studio 2008 project files.
Optional [arch] can be "Win64" or "IA64".
Visual Studio 8 2005 [arch] = Deprecated. Generates Visual Studio 2005
project files. Optional [arch] can be
"Win64".
Borland Makefiles = Generates Borland makefiles.
NMake Makefiles = Generates NMake makefiles.
NMake Makefiles JOM = Generates JOM makefiles.
Green Hills MULTI = Generates Green Hills MULTI files
(experimental, work-in-progress).
MSYS Makefiles = Generates MSYS makefiles.
MinGW Makefiles = Generates a make file for use with
mingw32-make.
Unix Makefiles = Generates standard UNIX makefiles.
Ninja = Generates build.ninja files.
Watcom WMake = Generates Watcom WMake makefiles.
CodeBlocks - MinGW Makefiles = Generates CodeBlocks project files.
CodeBlocks - NMake Makefiles = Generates CodeBlocks project files.
CodeBlocks - NMake Makefiles JOM
= Generates CodeBlocks project files.
CodeBlocks - Ninja = Generates CodeBlocks project files.
CodeBlocks - Unix Makefiles = Generates CodeBlocks project files.
CodeLite - MinGW Makefiles = Generates CodeLite project files.
CodeLite - NMake Makefiles = Generates CodeLite project files.
CodeLite - Ninja = Generates CodeLite project files.
CodeLite - Unix Makefiles = Generates CodeLite project files.
Sublime Text 2 - MinGW Makefiles
= Generates Sublime Text 2 project files.
Sublime Text 2 - NMake Makefiles
= Generates Sublime Text 2 project files.
Sublime Text 2 - Ninja = Generates Sublime Text 2 project files.
Sublime Text 2 - Unix Makefiles
= Generates Sublime Text 2 project files.
Kate - MinGW Makefiles = Generates Kate project files.
Kate - NMake Makefiles = Generates Kate project files.
Kate - Ninja = Generates Kate project files.
Kate - Unix Makefiles = Generates Kate project files.
Eclipse CDT4 - NMake Makefiles
= Generates Eclipse CDT 4.0 project files.
Eclipse CDT4 - MinGW Makefiles
= Generates Eclipse CDT 4.0 project files.
Eclipse CDT4 - Ninja = Generates Eclipse CDT 4.0 project files.
Eclipse CDT4 - Unix Makefiles= Generates Eclipse CDT 4.0 project files.

**Comment by NikolajBjorner** [comment]
This has all the traits of a 64 vs 32 bit incompatibility. Make sure that your vs command prompt is for the same architecture as the build configuration.

**Comment by delcypher** [comment]
@OisinMcArdle Have you tried using Z3's CMake build system. I advise avoiding `nmake`, it's really slow compared to something like CMake and [Ninja](https://ninja-build.org/).

**Comment by delcypher** [comment]
@OisinMcArdle You can find the docs [here](https://github.com/Z3Prover/z3/blob/master/README-CMake.md#ninja)

**Comment by OisinMcArdle** [comment]
@NikolajBjorner Thanks for your response. I'm sorry but can you explain how do I check that?
@delcypher Thanks for your response too. I'll start looking at this now

**Comment by delcypher** [comment]
@OisinMcArdle No problem. By the way the CMake build system doesn't build the java bindings by default so you'll need to pass `-DBUILD_JAVA_BINDINGS=ON` to the `cmake` invocation to configure the build. Make sure that the Z3 build architecture (i.e. x64 vs x86) matches your Java installation or you'll run into problems.

**Comment by OisinMcArdle** [comment]
@delcypher I've downloaded and installed cmake for windows. I've also gotten the ninja.exe. I don't mean to have you do most the work to help me do this, but I am not very familiar with VS command prompt and I haven't done something like this before. 
So checking the docs I need to remove the generated files from the python build which I've done by deleting the build folder.
I've created a new folder called build. I have the ninja.exe in there. 
So I just want to double check is this the correct command in case I create a bunch of useless files:
cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON

And sorry last question is where do I check the Z3 build architecture? Thanks.

**Comment by delcypher** [comment]
> So checking the docs I need to remove the generated files from the python build which I've done by deleting the build folder.

Unfortunately this is not enough. The old build system generates build files into the source tree (eurgh...). Run `git clean -fx src/` to fix this or just clone the git repo again.

> I've created a new folder called build. I have the ninja.exe in there.

You don't have to put `ninja.exe` in there. I would put somewhere more permanent and add that location to your `PATH` environment variable. 

So I just want to double check is this the correct command in case I create a bunch of useless files:

> cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON

That should be enough. The document I linked so lists the other options you can set.

> And sorry last question is where do I check the Z3 build architecture? Thanks.

CMake will detect what architecture MSVC is configured to use. If I recall correctly if you pass the `cl` command (the name of the MSVC compiler) for `/help` or `/version` (I can't remember which) it will report the architecture it will build for.

You can manually set up a developer command prompt by running the `vcvarsall.bat` script that ships with Visual Studio. It takes an argument that lets you specify the architecture to target. I think `x64` is `x86_64` (i.e. x86 64-bit) and `x86` is `i686` (i.e. x86 32-bit).

**Comment by OisinMcArdle** [comment]
@delcypher Thank you for all your help so far I think I'm on the last bit. 

- I just cloned the z3 again.
- I found the `vcvarsall.bat` script and and ran it with x64.
- I added the ninja.exe as a `PATH` variable
- I created build and `cd` into it.

When I ran the command:
`cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON`

I got:
`CMake Error: The source directory "C:/Z3Prover/z3/build" does not appear to contain CMakeLists.txt`

Looking into this do I have to create this file? Or is this supposed to be autogenearted?

**Comment by delcypher** [comment]
> I found the vcvarsall.bat script and and ran it with x64.

Okay. Given that you intend to use the Java bindings you should check that you have a 64-bit installation of Java (i.e. "windows x64" as listed at http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html )

> When I ran the command:
cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON
> 
> I got:
> CMake Error: The source directory "C:/Z3Prover/z3/build" does not appear to contain CMakeLists.txt

That's because your invocation of CMake is wrong. You need to tell it where the source tree is. CMake has no way of knowing where the source tree is because you can create the build directory anywhere on your system.

You should give it the path to the root of the Z3 git repository (i.e. the one that has on your system, e.g.

```
cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON ../
```

This assumes that you created the build inside the git repository.

> Update: @delcypher I copied the example\c files into the build directory.

That's completely wrong. Don't do that.

To fix what you've done just delete the build directory you created and start again.

```
# I'm assuming you are in the root of the Z3 git repository on your system
rmdir /q /s build

# Now We start again
mkdir build
cd build
cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON ../
# Assuming everything went right we can now build
ninja
```

**Comment by delcypher** [comment]
Another note is CMake is picking up an installed copy of gcc on your system

```
-- The C compiler identification is GNU 5.2.0
-- The CXX compiler identification is GNU 5.2.0
-- Check for working C compiler: C:/Program Files/Haskell Platform/7.10.3/mingw/bin/gcc.exe
-- Check for working C compiler: C:/Program Files/Haskell Platform/7.10.3/mingw/bin/gcc.exe -- works
```

Do you actually want that? If not you might need to be explicit to CMake that you want to use MSVC. If you already have a build directory configured to use gcc you will need to delete it because once CMake selects the compiler you cannot change it for that build directory.

```
SET CXX=cl
cmake -G "Ninja" -DBUILD_JAVA_BINDINGS=ON ../
```

I can't guarantee the above will work though. Your windows set up is different from mine so you will need to figure out how to set up your environment correctly.

**Comment by NikolajBjorner** [comment]
Can we close this?

For the python build, you seem to want to build x64, so you need to pass -x to the build script per the documentation.

C:\z3\build>..\scripts\mk_make.py -h
mk_make.py: Z3 Makefile generator

This script generates the Makefile for the Z3 theorem prover.
It must be executed from the Z3 root directory.

Options:
  -h, --help                    display this message.
  -s, --silent                  do not print verbose messages.
  --parallel=num                use cl option /MP with 'num' parallel processes
  --pypkgdir=<dir>              Force a particular Python package directory (default C:\Python27\Lib\site-packages)
  -b <subdir>, --build=<subdir>  subdirectory where Z3 will be built (default: build).
  --githash=hash                include the given hash in the binaries.
  --git-describe                include the output of 'git describe' in the version information.
  -d, --debug                   compile Z3 in debug mode.
  -t, --trace                   enable tracing in release mode.
  --guardcf                     enable Control Flow Guard runtime checks.
  **-x, --x64                     create 64 binary when using Visual Studio.**

**Comment by suhaorhun** [comment]
Hi,

Sorry to revive a closed issue, but I encountered exactly the same error while trying to install Z3. In my case the problem is due to working on a regular command line which was opened when I follow Tools > Visual Studio Command Prompt inside the VS. I think it opens a 32 bit command line by default. The problem disappeared when I opened "x64 Native Tools Command Prompt for VS 2017" and ran nmake on it.

I wanted to leave the solution here in case someone else encounters the same problem for the same reason and ends up here.

By the way, thanks a lot for the great work on Z3.

Best,
Suha
